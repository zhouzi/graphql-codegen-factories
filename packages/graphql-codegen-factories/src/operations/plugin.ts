import { Kind, FragmentDefinitionNode, concatAST, DocumentNode } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesOperationsVisitor,
  FactoriesOperationsVisitorRawConfig,
} from "./FactoriesOperationsVisitor";

export const plugin: PluginFunction<
  FactoriesOperationsVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const allAst = concatAST(
    documents.map(({ document }) => document as DocumentNode)
  );
  const fragments = allAst.definitions.filter(
    (d) => d.kind === Kind.FRAGMENT_DEFINITION
  ) as FragmentDefinitionNode[];
  const allFragments: FragmentDefinitionNode[] = [
    ...fragments,
    // `externalFragments` config is passed by the near-operation-file preset.
    // It is an array of fragments declared outside of the operation file.
    ...(config.externalFragments || []).map(({ node }) => node),
  ];

  const visitor = new FactoriesOperationsVisitor(schema, allFragments, config);
  const content = (
    oldVisit(allAst, { leave: visitor }) as DocumentNode
  ).definitions
    .filter((definition) => typeof definition === "string")
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
