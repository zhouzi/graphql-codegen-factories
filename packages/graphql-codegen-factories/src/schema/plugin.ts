import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import { FactoriesBaseVisitorRawConfig } from "../FactoriesBaseVisitor";
import { FactoriesSchemaVisitor } from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesBaseVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);

  const visitor = new FactoriesSchemaVisitor(schema, config);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
