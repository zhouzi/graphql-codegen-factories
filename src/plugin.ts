import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesVisitor,
  FactoriesVisitorRawConfig,
} from "./FactoriesVisitor";

export const plugin: PluginFunction<
  FactoriesVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.config.typesPath
      ? [
          `import * as ${visitor.config.namespacedImportName} from '${visitor.config.typesPath}';\n`,
        ]
      : [],
    content,
  };
};
