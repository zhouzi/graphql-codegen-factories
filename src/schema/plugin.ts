import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesSchemaVisitor,
  FactoriesSchemaVisitorRawConfig,
} from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesSchemaVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesSchemaVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
