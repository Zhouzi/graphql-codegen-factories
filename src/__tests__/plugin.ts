import { buildSchema } from "graphql";
import { plugin } from "../plugin";

describe("plugin", () => {
  it("should create factories with built-in types", () => {
    const schema = buildSchema(`
      type User {
        id: ID!
        organizationId: ID
        email: String!
        name: String
        age: Int!
        followers: Int
        geo: Geo!
        isUser: Boolean!
        isAdmin: Boolean
        status: UserStatus!
        favouriteFruit: Fruit
        posts: [Post!]!
        subscribers: [User!]
      }

      type Geo {
        lat: Float!
        lon: Float
      }

      type Post {
        id: ID!
      }

      enum UserStatus {
        Activated
        Created
      }

      enum Fruit {
        Pineapple
        Mango
      }
    `);
    expect(plugin(schema, [], {})).toMatchSnapshot();
  });

  it("should use enums as types", () => {
    const schema = buildSchema(`
      type User {
        status: UserStatus!
      }

      enum UserStatus {
        Activated
        Created
      }
    `);
    expect(plugin(schema, [], { enumsAsTypes: true })).toMatchSnapshot();
  });

  it("should use the custom scalar defaults", () => {
    const schema = buildSchema(`
      type User {
        createdAt: Date!
      }
      
      scalar Date
    `);
    expect(
      plugin(schema, [], {
        scalarDefaults: {
          Date: "new Date()",
        },
      })
    ).toMatchSnapshot();
  });

  it("should create factories for inputs", () => {
    const schema = buildSchema(`
      input PostInput {
        id: ID
        title: String!
      }
    `);
    expect(plugin(schema, [], {})).toMatchSnapshot();
  });

  it("should not create factories for Query and Mutation", () => {
    const schema = buildSchema(`
      type User {
        id: ID!
      }

      type Query {
        users: [User!]!
      }

      type Mutation {
        createUser(id: ID!): User!
      }
    `);
    expect(plugin(schema, [], {})).toMatchSnapshot();
  });

  it("should customize the factory name", () => {
    const schema = buildSchema(`
      type User {
        id: ID!
      }
    `);
    expect(
      plugin(schema, [], {
        factoryName: "new{Type}",
      })
    ).toMatchSnapshot();
  });
});
