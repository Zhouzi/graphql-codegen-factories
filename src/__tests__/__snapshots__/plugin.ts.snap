// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should create factories for inputs 1`] = `
"export function createPostInputMock(props: Partial<PostInput>): PostInput {
  return {
    id: null,
    title: \\"\\",
    ...props,
  };
}
"
`;

exports[`plugin should create factories with built-in types 1`] = `
"export function createUserMock(props: Partial<User>): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    organizationId: null,
    email: \\"\\",
    name: null,
    age: 0,
    followers: null,
    geo: createGeoMock({}),
    isUser: false,
    isAdmin: null,
    status: UserStatus.Activated,
    favouriteFruit: null,
    posts: [],
    subscribers: null,
    ...props,
  };
}

export function createGeoMock(props: Partial<Geo>): Geo {
  return {
    __typename: \\"Geo\\",
    lat: 0,
    lon: null,
    ...props,
  };
}

export function createPostMock(props: Partial<Post>): Post {
  return {
    __typename: \\"Post\\",
    id: \\"\\",
    ...props,
  };
}
"
`;

exports[`plugin should customize the factory name 1`] = `
"export function newUser(props: Partial<User>): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    ...props,
  };
}
"
`;

exports[`plugin should not create factories for Query and Mutation 1`] = `
"export function createUserMock(props: Partial<User>): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    ...props,
  };
}
"
`;

exports[`plugin should use enums as types 1`] = `
"export function createUserMock(props: Partial<User>): User {
  return {
    __typename: \\"User\\",
    status: \\"Activated\\",
    ...props,
  };
}
"
`;

exports[`plugin should use the custom scalar defaults 1`] = `
"export function createUserMock(props: Partial<User>): User {
  return {
    __typename: \\"User\\",
    createdAt: new Date(),
    ...props,
  };
}
"
`;
